Tasks/AzureVmssDeployment/operations/VirtualMachineScaleSet.ts

```diff
@@ -24,135 +25,143 @@ export default class VirtualMachineScaleSet {
-                                    "commandToExecute": this.taskParameters.customScriptCommand
+                        var extensionMetadata: azureModel.VMExtensionMetadata = null;
+                        var customScriptExtension: azureModel.VMExtension = null;
+                        if(!!this.taskParameters.customScriptUrl) {
+                            extensionMetadata = this._getCustomScriptExtensionMetadata(osType);
+                            customScriptExtension = {
+                                name: "CustomScriptExtension" + Date.now().toString(),
+                                properties: {
+                                    type: extensionMetadata.type,
+                                    publisher: extensionMetadata.publisher,
+                                    typeHandlerVersion: extensionMetadata.typeHandlerVersion,
+                                    autoUpgradeMinorVersion: true,
+                                    settings: {
+                                        "fileUris": [ this.taskParameters.customScriptUrl ]
+                                    },
+                                    protectedSettings: {
+                                        "commandToExecute": this.taskParameters.customScriptCommand
+                                    }
                                 }
-                            }
-                        };
+                            };
 
-                        this._getExistingCustomScriptExtension(client, resourceGroupName, customScriptExtension, (error, matchingExtension, request, response) => {
+                            var matchingExtension = await this._getExistingCustomScriptExtension(client, resourceGroupName, customScriptExtension);
                             // if extension already exists, remove it
                             if(!!matchingExtension) {
-                                this._deleteAndInstallCustomScriptExtension(client, resourceGroupName, matchingExtension, customScriptExtension, resolve, reject, (error, matchingExtension, request, response) => {
-                                    this._updateImageInternal(client, resourceGroupName, resolve, reject);
-                                });
+                                await this._deleteAndInstallCustomScriptExtension(client, resourceGroupName, matchingExtension, customScriptExtension);
+                                await this._updateImageInternal(client, resourceGroupName);
                             } else {
-                                this._installCustomScripExtension(client, resourceGroupName, customScriptExtension, resolve, reject, (error, result, request, response) => {
-                                    this._updateImageInternal(client, resourceGroupName, resolve, reject);
-                                });
+                                await this._installCustomScripExtension(client, resourceGroupName, customScriptExtension);
```

Comment:
there were few reasons why I did like this:
1. deleting & installing extension before image update will not cause any downtime. The custom script is supposed to run at start-up only. Removing extension is not going to have any adverse affect.
2. I was seeing some reliability issues when earlier I was doing both in same PATCH call. it used to hang sometimes.
3. Discussed with Roopesh and this looks better way to achieve blue-green deployment. A cloned VMSS will be used to update extension and image and then vip swap will be done. So no downtime.
4. Anyways we have to support setting extension as a separate functionality. This makes it re-usable to enable that easily.